### ISSUE TITLE
🐞 CLI: dagger listen --disable-host-read-write fails in v0.9.9

### ISSUE THREAD

**Issue by matipan** [dagger/dagger#6640]
### What is the issue?

There is a regression on dagger v0.9.9 with the dagger listen command when the flag `--disable-host-read-write` is specified. Prior to v0.9.8 this was working correctly. Since v0.9.9 it fails with:

```
Error: failed to get configured module: failed to get local root path: input: resolve: moduleSource: resolveContextPathFromCaller: failed to stat source root: failed to receive file bytes message: rpc error: code = Unimplemented desc = unknown service moby.filesync.v1.FileSync
```

If you run this command inside a module it also fails with the same error:

```
$ dagger mod init --name test --sdk go
$ dagger listen --disable-host-read-write
├ [0.01s] loading module
✘ directory ERROR [0.01s]
▶ directory ▶ host.directory /home/matipan/bin/test
  ✘ upload /home/matipan/bin/test from pop-os (client id: 09oib86lak5rdy6j1tbxf2gbu) ERROR [0.01s]
  ├ [0.00s] transferring /home/matipan/bin/test:
• Engine: 6241366fb45d (version v0.9.7)
⧗ 1.97s ✔ 6 ✘ 3
Error: failed to get loaded module ID: input: resolve: host: directory: host directory /home/matipan/bin/test: no local sources enabled
input: resolve: host: directory: host directory /home/matipan/bin/test: no local sources enabled
```

### Dagger version

dagger v0.9.9 ([registry.dagger.io/engine](http://registry.dagger.io/engine)) linux/amd64

### Steps to reproduce

```
$ dagger listen --progress plain --disable-host-read-write
```

### Log output

```
Connected to engine 5c478db0e017 (version v0.9.9)

Error: failed to get configured module: failed to get local root path: input: resolve: moduleSource: resolveContextPathFromCaller: failed to stat source root: failed to receive file bytes message: rpc error: code = Unimplemented desc = unknown service moby.filesync.v1.FileSync
```

**Comment by jedevc** [comment]
This is likely caused by the changes to `FileSync` in https://github.com/dagger/dagger/pull/6575.

**Comment by gerhard** [comment]
I am wondering if instead of fixing this, we should stop using `listen` altogether in the Playground, and maybe remove this feature. `dagger listen` came up a few times times in the past, and I'm still not sure why we have it:
- https://github.com/dagger/dagger/issues/4319
- https://github.com/dagger/dagger/issues/5205

The last thread is particularly interesting, since it offers a bunch of alternatives to `dagger listen`. As that conversation seems to be long-term and filled with bike shedding, I would rather focus on something pragmatic for the Playground.

As we are consolidating everything on EKS, will `dagger listen` still make sense? I am thinking:
- A dedicated node pool for Playground
- Playground Engines will become just a DaemonSet
- One of the containers in that DaemonSet will be the API
  - This ensures that both change as a single unit
  - Liveness checks will ensure that the DaemonSet will not be rolled if there is some sort of incompatibility between the API and the Engine (they are a single unit)
  - Alternatively, we do not do in-place updates, but deploy new versions in parallel (I'm personally a big fan of this)
- API container connects to the Dagger Engine via a unix socket, no TCP port (they are a single unit)
- The Web instance selects which version gets used by passing an `X-Dagger-Version` header - nice trick @matipan 🙌

If we went in the direction described above, would we need `dagger listen` @marcosnils ?

**Comment by helderco** [comment]
I use dagger listen sometimes to inspect the API with [https://studio.apollographql.com/](https://studio.apollographql.com/), for example. I'd be happy with a command that can spin up a playground locally though.

**Comment by jedevc** [comment]
Sounds like a legit reason to keep `listen` around to me! I actually wasn't aware of it, might have to start using it now.

That said, if we do keep it, it definitely needs a bit of love - including some tests and docs.

**Comment by matipan** [comment]
I've done a bit of a deep dive into what might be causing this issue. 

**TL;DR**: in #6575 we started to resolve `LocalContextPath` in the context of the Dagger Engine instead of the host, which requires communicating with the host through gRPC. However, the required services are disabled due to `--disable-host-read-write` so `LocalContextPath` cannot be resolved.

Tracking from where the error originates we start in `StatCallerHostPath`:
```go
func (c *Client) StatCallerHostPath(ctx context.Context, path string, returnAbsPath bool) (*fsutiltypes.Stat, error) {
        // ...
	diffCopyClient, err := filesync.NewFileSyncClient(clientCaller.Conn()).DiffCopy(ctx) // <-- DiffCopy is not available
	if err != nil {
		return nil, fmt.Errorf("failed to create diff copy client: %s", err)
	}
	defer diffCopyClient.CloseSend()
	msg := fsutiltypes.Stat{}
	err = diffCopyClient.RecvMsg(&msg)
	if err != nil {
		return nil, fmt.Errorf("failed to receive file bytes message: %s", err) // <--- This is where it errors out
	}
	return &msg, nil
}
```

`StatCallerHostPath` is used by `resolveContextPathFromCaller` to effectively resolve the `LocalContextPath` of the module when we are initializing the configuration of a module on all commands (done by `optionalModCmdWrapper`). This functions are all executed in the context of the Dagger Engine. But since they are trying to resolve paths that are specific to the host they need to make gRPC calls to the buildkit client running in the host to traverse and copy the contents of some directories. However, when `DisableHostRW` is enabled those services do not get registered:
```go
// filesync
if !c.DisableHostRW {
    bkSession.Allow(AnyDirSource{})
    bkSession.Allow(AnyDirTarget{})
}
```

How did this used to work in v0.9.8? It seems that both the `LocalContextPath` (previously called `LocalRootPath`) and `LocalRootSourcePath` (previously called `LocalSourcePath`) were resolved in the context of the CLI. Meaning that no path resolution ever exited the host, so having the FileSync `DiffCopy` API disabled had no effect.

I'm sure there are good reasons for it, but by just looking at the code I'm not able to understand why `LocalContextPath` is now resolved by calling the engine. Doing so is what essentially breaks this particular behavior.

Given that I lack this context, I'm not entirely certain what the most appropriate solution would be. If going back to resolve `LocalContextPath` without exiting the host, adding a fallback to use the host when `DisableHostRW` is enabled or some other alternative. 

I'm happy to keep going, but I think it's worth checking if this all makes sense @jedevc @sipsma

**Comment by vito** [comment]
Quick drive-by comment: I think if `--disable-host-read-write` is passed we can probably just skip any module loading related code. We only use that flag for Playground, which doesn't use modules.

Related, maybe there's a work-around where you can pass `-m github.com/foo/bar` (literally any module)? That way it'll load a remote module instead of trying to figure out if you're running from a local one.

**Comment by sipsma** [comment]
Yeah for more context, module loading all happens from the engine now because:
1. It centralized the logic more (module loading previously was split between the cli and engine, now there's only one place to keep track of it all
2. The module loading logic got way more complex recently due to the need to sparsely load local modules (which multiplies the above point)

The behavior of daggery query is supposed to be that it tries to load a local module if it exists but ignores any error and just operates without the module if it can't load one. I suspect that we just need to ignore an error somewhere. I will take a look at this, assigning to me.
